utils/estimated_histogram.hh

```diff
@@ -47,13 +47,376 @@
+     * When entering a value:
+     * All values lower than the MIN will be included in the first bucket.
+     * All values higher than MAX will be included the last bucket that serves as the
+     * infinity bucket.
+     *
+     * buckets are distributed as pseudo floating point:
+     * The range [MIN, MAX) is split into log2 ranges.
+     *     ranges = log2(max/min)
+     * Each of that ranges is split according to the number of buckets:
+     *     resolution = (NUM_BUCKETS - 1)/ranges
+     *
+     * For example, if the MIN value is 128, the MAX is 1024 and the number of buckets is 13:
+     *
+     * Anything below 128 will be in the bucket 0, anything above 1024 will be in bucket 13.
+     *
+     * the range [128, 1024) will be split into log2(1024/128) = 3:
+     * 128, 256, 512, 1024
+     *
+     * Each range is split into 12/3 = 4.
+     * 128            | 256            | 512            | 1024
+     * 128 160 192 224| 256 320 384 448| 512 640 768 896|
+     *
+     *
+     * Calculating the bucket limit of bucket i:
+     * The range: 2^(i/4)* min
+     * The sub range: i%4 * range/4
+     *
+     * How to find a bucket index for a value.
+     * The bucket index consist of two part:
+     * higher bits are based on log2(value/min)
+     *
+     * lower bits are based on the high 2 MSB (ignoring the leading 1).
+
+     * for example: 330 (101001010)
+     * higher bits: log2(330/128) = 1
+     * low bit MSB: 330 = 01 (the lower two bits out of the upper 3)
+     * So the index: 101 = 5
+     *
+     *
+     * About the min/max and number of buckets.
+     * ========================================
+     *
+     * For MIN and MAX choose numbers that are a power of 2.
+     * The number of buckets will determine the resolution that should also be a power of 2
+     * So the total number of bucket should be log2(MAX/MIN) * Resolution + 1
+     *
+     * Limitation: You must set the MIN value to be higher then the resolution.
+     * For example, for a 2 bits resolution MIN should be 2^2 = 4 or higher.
+     *
+     */
+    using histogram_type = pseudo_floatingpoint_histogram<MIN, MAX, NUM_BUCKETS>;
+
+    static constexpr uint8_t RANGES = pow2_rank_constexpr(MAX/MIN);
+    static constexpr uint8_t RESOLUTION = (NUM_BUCKETS - 1)/RANGES;
+    static constexpr uint8_t RESOLUTION_BITS = pow2_rank_constexpr(RESOLUTION);
+    static constexpr uint8_t BASESHIFT = (MIN == 0) ? 0 : pow2_rank_constexpr(MIN);
+    static constexpr uint64_t LOWER_BITS_MASK = (1 << RESOLUTION_BITS) - 1;
+
+    std::vector<uint64_t> buckets;
+
+    pseudo_floatingpoint_histogram() {
+        assert(BASESHIFT >= RESOLUTION);
+        buckets.resize(NUM_BUCKETS, 0);
+    }
+
+    /*!
+     * \brief Returns the bucket limit given the bucket id.
+     * The first and last bucket will always return the MIN and MAX respectively.
+     *
+     */
+    uint64_t get_bucket_limit(uint16_t bucket_id) const {
```

Comment:
Please make it explicit in the name whether this is the upper bound or lower bound.